internal/app/starter/master_linux.go

```diff
@@ -20,93 +20,119 @@ import (
-				return
-			}
-			if err := obj.PreStartProcess(containerPid, conn, fatalChan); err != nil {
-				fatalChan <- fmt.Errorf("pre start process failed: %s", err)
+
+			if err != io.EOF {
+				fatalChan <- fmt.Errorf("error while reading master socket data: %s", err)
 				return
 			}
-		}
-		// wait container process execution, EOF means container process
-		// was executed and master socket was closed by stage 2. If data
-		// byte sent is equal to 'f', it means an error occurred in
-		// StartProcess, just return by waiting error and process status
-		_, err = conn.Read(data)
-		if (err != nil && err != io.EOF) || data[0] == 'f' {
+			// EOF means something goes wrong in stage 2, don't send error via
+			// fatalChan, error will be reported by stage 2 and the process
+			// status will be set accordingly via MonitorContainer method below
+			sylog.Debugf("stage 2 process was interrupted, waiting status")
+			return
+		} else if data[0] == 'f' {
+			// StartProcess reported an error in stage 2, don't send error via
+			// fatalChan, error will be reported by stage 2 and the process
+			// status will be set accordingly via MonitorContainer method below
 			sylog.Debugf("stage 2 process reported an error, waiting status")
 			return
 		}
-
-		err = engine.PostStartProcess(containerPid)
-		if err != nil {
-			fatalChan <- fmt.Errorf("post start process failed: %s", err)
+		if err := obj.PreStartProcess(containerPid, conn, fatalChan); err != nil {
+			fatalChan <- fmt.Errorf("pre start process failed: %s", err)
 			return
 		}
-		if isInstance && os.Getppid() == ppid {
-			// sleep a bit to see if child exit
-			time.Sleep(100 * time.Millisecond)
-			syscall.Kill(ppid, syscall.SIGUSR1)
-		}
+	}
+	// wait container process execution, EOF means container process
+	// was executed and master socket was closed by stage 2. If data
+	// byte sent is equal to 'f', it means an error occurred in
+	// StartProcess, just return by waiting error and process status
+	_, err = conn.Read(data)
+	if (err != nil && err != io.EOF) || data[0] == 'f' {
+		sylog.Debugf("stage 2 process reported an error, waiting status")
+		return
+	}
+
+	err = engine.PostStartProcess(containerPid)
+	if err != nil {
+		fatalChan <- fmt.Errorf("post start process failed: %s", err)
+		return
+	}
+	if isInstance && os.Getppid() == ppid {
+		// sleep a bit to see if child exit
+		time.Sleep(100 * time.Millisecond)
+
+		syscall.Kill(ppid, syscall.SIGUSR1)
+	}
+}
+
+// Master initializes a runtime engine and runs it
+func Master(rpcSocket, masterSocket int, isInstance bool, containerPid int, engine *engines.Engine) {
+	var fatal error
+	var status syscall.WaitStatus
+
+	fatalChan := make(chan error, 1)
+
+	if engine == nil || containerPid < 0 {
+		fatalChan <- fmt.Errorf("invalid parameter(s)")
+		return
+	}
+
+	ppid := os.Getppid()
+
+	go func() {
+		doRPCthread(rpcSocket, containerPid, engine, fatalChan)
+		runtime.Goexit()
```

Comment:
what's the logic of keeping this here vs in the doRPCthread function? I can guess, but I would prefer to have it documented so that people looking at this in the future don't have to.